c7n/tags.py

```diff
@@ -917,6 +919,97 @@ def process_resource_set(self, resource_set, tags):
+
+        policies:
+          - name: copy-tags-from-ebs-volume-to-snapshot
+            resource: ebs-snapshot
+            actions:
+              - type: copy-related-resource-tag
+                resource: ebs
+                skip_missing: True
+                key: VolumeId
+                copy_tags:
+                  - *
+    """
+
+    schema = utils.type_schema(
+        'copy-related-resource-tag',
+        resource={'type': 'string'},
+        skip_missing={'type': 'boolean'},
+        key={'type': 'string'},
+        copy_tags={'type': 'array'},
+    )
+    permissions = ()
+
+    def validate(self):
+        related_resource = self.data['resource']
+        resource_manager = self.manager.get_resource_manager(related_resource)
+        if resource_manager is {}:
+            raise PolicyValidationError(
+                "Error: Invalid resource type selected: %s" % related_resource
+            )
+        if self.manager.action_registry.get('tag') is None:
+            raise PolicyValidationError(
+                "Error: Tag action missing on resource"
+            )
+
+    def process(self, resources):
+        tag_keys = self.data['copy_tags']
+        related_key = self.data['key']
+        related_type = self.data['resource']
+
+        related_resource_tuple = list(
+            zip(jmespath.search('[].%s' % related_key, resources), resources))
+
+        related_ids = [r[0] for r in related_resource_tuple]
+        related_tag_map = self.get_resource_tag_map(related_type, related_ids)
+        missing_related_tags = list(set(related_ids).difference(related_tag_map.keys()))
+
+        if not self.data.get('skip_missing', True) and missing_related_tags:
+            raise PolicyExecutionError(
+                "Unable to find all %s resources associated with %s" % (related_type, related_ids))
+
+        for related, r in related_resource_tuple:
+            if related in missing_related_tags:
+                self.log.warning(
+                    'Tags not found for related resource: %s, skipping %s' % (related, r))
+                continue
+            related_tags = related_tag_map[related]
+            self.process_resource(r, related_tags, tag_keys)
+
+    def process_resource(self, r, related_tags, tag_keys):
+        related_resource_tags = {t['Key']: t['Value'] for t in related_tags}
+        add_tags = {}
+        if '*' in tag_keys:
+            add_tags = related_resource_tags
+        else:
+            for t in tag_keys:
+                if t in related_resource_tags:
+                    add_tags[t] = related_resource_tags[t]
+        tagger = copy.deepcopy(self)
+        tagger['tags'] = add_tags
+        # rely on resource manager tag action implementation as it can differ between resources
+        tag_action = self.manager.action_registry.get('tag')
+        self.log.info('Adding related tags: %s to resource: %s' % (add_tags, r))
+        tag_action.process(tagger, resources=[r])
```

Comment:
i would consider this a hot path, and would recommend reusing objects, rather than deep copying per resource.